### ISSUE TITLE
Discrepancy in type disambiguation between `Pstr_value` and `Pexp_let` 

### ISSUE THREAD

**Issue by johnyob** [ocaml/ocaml#11462]
Hey 👋. For the following module:
```ocaml
module M = struct
  type t = X of { x : int }

  let x = X { x = 1 }
end;;
```
The top-level definition 
```ocaml
let (X { x }) = M.x;;
``` 
yields the type error: ```Error: Unbound constructor X```.  However, no error occurs w/ 
```ocaml
let x = 
  let (X { x }) = M.x in
  x
;;
val x : int = 1
```
Is this a bug?

Credit to @EduardoRFS for finding this.

**Comment by gasche** [comment]
It's probably a bug (in the sense that I would also expect the two programs to behave in the same way), but we can provide a reasonable explanation for what is going on. Local let-bindings `let (K p) = e1 in e2`, where `K` is a variant constructor, are actually translated into `match e1 with K p -> e2` before type-checking. `match` and toplevel-`let` use different disambiguation rules, which explains the discrepancy here. There is no such transformation on toplevel lets, because there isn't a well-defined body `e2` to use as the right-hand-side, but you could manually write `let x = match M.x with X { x } -> x` which is accepted.

**Comment by Octachron** [comment]
I agree that it might be worth it to harmonize the behavior of the type-directed disambiguation in both cases.
